algorithms/AR_tag.py

```diff
@@ -0,0 +1,99 @@
+from collections import namedtuple
+from math import dist
+import cv2
+import logging
+from numpy.core.numeric import NaN
+import itertools
+import core
+import numpy as np
+
+tag_cascade = cv2.CascadeClassifier("resources/tag_detection/cascade30.xml")
+
+Tag = namedtuple("Tag", ["cX", "cY", "distance", "angle"])
+
+
+def detect_ar_tag(reg_img):
+    """
+    Detects an AR Tag in the provided color image.
+
+    Parameters:
+    -----------
+        reg_img - the provided image we are looking at to find an ar tag
+
+    Returns:
+    --------
+        tags - a list of Tags (named tuple) that contain the (cX, cY, distance, angle) of the detected AR tags
+        reg_img - the image with detected AR Tags drawn on top of it
+    """
+
+    gray = cv2.cvtColor(reg_img, cv2.COLOR_BGR2GRAY)
+    tags_coords = tag_cascade.detectMultiScale(gray, 1.3, 5)
+
+    tags = []
+    for (x, y, w, h) in tags_coords:
+        reg_img = cv2.rectangle(reg_img, (x, y), (x + w, y + h), (255, 0, 0), 2)
+
+        # Calculate the center points of the AR Tag
+        cX = x + (w / 2)
+        cY = y + (h / 2)
+
+        # Find the distance/angle of said center pixels
+        distance, angle = track_ar_tag((cX, cY))
+        tags.append(Tag(cX, cY, distance, angle))
+
+    return tags, reg_img
+
+
+def track_ar_tag(center):
+    """
+    Track the distance and angle of the AR Tag from the perspective of the Rover.
+
+    Parameters:
+    -----------
+        center - the X, Y of the center pixels of the AR Tag
+
+    Returns:
+    --------
+        distance - the distance in meters to the AR Tag
+        angle - the angle in degrees from the rover to the AR Tag. Left is negative,
+        right is positive
+    """
+
+    logger = logging.getLogger(__name__)
+
+    # Center coordinates
+    cX, cY = center
+
+    # Depth image is at half resolution
+    cX = int(cX / 2)
+    cY = int(cY / 2)
+
+    # Grab the distance from the depth map
+    distance = NaN
+
+    # Find some permutations we can use in case of noisy data
+    coordinates = [0, 1, -1, 2, -2]
+    perm = list(itertools.permutations(coordinates, 2))
+
+    # Grab the distance from the depth map, iterating over pixels if the distance is not finite
+    index = 0
+
+    while not np.isfinite(distance):
+        if index < len(perm):
+            distance = core.vision.camera_handler.grab_depth_data()[cY + perm[index][1]][cX + perm[index][0]]
+            index += 1
+        else:
+            index = 0
```

Comment:
At first glance this loop seems like it may be able to "infinite" loop. Index is going to increase and loop through all the pixels in perm, and if it gets to the end it is just going to start over? Is the thought here that instead of checking all pixels in perm and then breaking out if all are NaN, we just keep going and waiting until the camera feed updates and hopefully then one of them will not be NaN? I am a little unclear how this fits in to the bigger picture and doesn't prevent the work of num_detection_attempts

Reply:
Yeah, so this is a hack that probably originated from filming. It's kind of a tricky situation in where we need to estimate the position of the AR Tag once we have found it, but we don't always have a viable depth estimate for it. I fixed it by making it wait until we did, because it should pretty much always work.

This differs from num_detection_attempts because this is when we have already IDed an AR Tag and are now trying to find it's real world coordinates. Num_detection_attempts is the number of times we try to identify an AR Tag before we accept we lost sight.

Now of course there are edge cases, so I'll take a look and might move this into an issue, or fix it if I find a nice quick solution.